### PR TITLE
Add fake process group

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102180

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/testing/_internal/distributed/fake_pg.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# A fake process group (not related to FakeTensor) is a process group which
+# doesn't actually do any communication, it just hallucinates some
+# communication.  You can run a single rank with a fake process group
+# without needing multiple processes.
+
+class FakeProcessGroup(dist.ProcessGroup):
+    pass
+
+class FakeStore(dist.Store):
```
Comment (wanchaol) [inline torch/testing/_internal/distributed/fake_pg.py:11]: we probably don't need a FakeStore and instead we can just use HashStore I suppose. But that would be in a separate PR.

**Review by wanchaol — APPROVED** [review]
looks great, thanks!

File: test/distributed/test_fake_pg.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+from torch.testing._internal.common_utils import (
+    TestCase,
+    run_tests,
+)
+
+HAS_CUDA = torch.cuda.is_available()
+
+class Test(TestCase):
+    @unittest.skipIf(not HAS_CUDA, 'No CUDA')
```
Comment (wanchaol) [inline test/distributed/test_fake_pg.py:23]: Looks like we should import unittest or add `skipIf` to the import to silent CI

File: torch/testing/_internal/distributed/fake_pg.py (hunk 1 above)
Comment (wanchaol) [reply torch/testing/_internal/distributed/fake_pg.py:11]: Hmmm I think FakePG barrier should be a no-op, and in terms of `init_processs_group` store based barrier it seems we already skip the barrier so it won't write anything to the HashStore. So either fake_store or hash_store could work I feel (I can give it a try and see if that's feasible or not)

Comment (ezyang) [reply torch/testing/_internal/distributed/fake_pg.py:11]: FSDP does a barrier which is why I ended up doing FakeStore. But yeah, try some stuff out, the goal is to be able to run FSDP end-to-end with the fake group with only one node.

**Review by albanD — COMMENTED** [review]
Very cool!
Do you have a small script showing the right way to use this?

**Comment by ezyang** [comment]
the test cases are pretty good

**Comment by albanD** [comment]
Ok!
But can I actually forward/backward with that Module? What should I expect the memory behavior of this thing to be (representative of what would happen with nccl or not?)

**Comment by wanchaol** [comment]
> Ok! But can I actually forward/backward with that Module? What should I expect the memory behavior of this thing to be (representative of what would happen with nccl or not?)

@albanD the memory behavior should be representative of each rank behavior as if you are running a real multiprocessing job with nccl process group. i.e. if specify `rank=1` and `torch.cuda.set_device(1)` when you initialize the fake pg, all of the memory behaviors should be similar as if you are in `rank 1` of a real multiprocessing job. (as under the hood each rank still allocate the same amount of data for c10d collectivese)

**Comment by insujang** [comment]
May I ask why you implemented additional `FakeProcessGroup` implementation while we have `MockProcessGroup` in `torch/testing/_internal/distributed/distributed_utils.py`? I understand that `MockProcessGroup` is just mocking a process group and does nothing while `FakeProcessGroup` aims to provide fake communication as well, but it is a superset and seems no reason to maintain `MockProcessGroup`. Will you deprecate `MockProcessGroup` later?
